vault/expiration.go

```diff
@@ -1785,9 +1792,11 @@ func (m *ExpirationManager) lookupLeasesByToken(ctx context.Context, te *logical
 
 // emitMetrics is invoked periodically to emit statistics
 func (m *ExpirationManager) emitMetrics() {
-	m.pendingLock.RLock()
-	num := len(m.pending)
-	m.pendingLock.RUnlock()
+	num := 0
+	m.pending.Range(func(_, _ interface{}) bool {
```

Comment:
This is the only part that gives me pause.  I'd suggest replacing it with a standalone atomic that we inc/dec as we mutate the map, but having tried that for something else I'm working on, it's trickier than you might think.  (My counter wound up going negative for reasons I'm still making sense of).

Reply:
In case it's not clear, my concern is that iterating over all leases every second may be a lot of CPU at customers with big lease sets.

Re the problems with the standalone atomic gauge approach: I think the issues I ran into are because when we call stopExpiration, we wipe the pending map and set Core.expiration to nil, but we don't unlink the token store via SetExpirationManager(nil), and there may be in-flight operations that already have a reference to c.expiration to proceed.  I'm sort of thinking we shouldn't, that there should be a new mode used once stopExpiration has been called, which causes all ExpirationManager methods to fail without doing anything.

Since that would be a 4th mode for ExpirationManager (along with restore, tidy, and normal) we should consider introducing a state machine if we add this.

Reply:
Hm, I didn't realize this was 1/sec.

With the new approach we don't swap out the pending map so it may be easier to update the counter only when pendingLock is held.

Reply:
Oh good point, that may help.  Are you going to explore that in this PR, or would you rather move ahead with the current implementation for now?

Reply:
I submitted a change that includes a count.

I was wondering whether to make a wrapper around sync.Map instead, but... there's one place where we rely on the contextual knowledge that a particular key exists, which would be difficult to encapsulate without re-doing the lookup.